jobs/webcompat-kb/webcompat_kb/main.py

```diff
@@ -590,30 +672,140 @@ def filter_bug_history_changes(self, history):
+                    if bug_id not in keyword_history:
+                        keyword_history[bug_id] = {"added": set(), "removed": set()}
+                    if change["added"]:
+                        keyword_history[bug_id]["added"].update(
+                            change["added"].split(", ")
+                        )
+                    if change["removed"]:
+                        keyword_history[bug_id]["removed"].update(
+                            change["removed"].split(", ")
+                        )
+        return keyword_history
+
+    def create_synthetic_history(self, bugs, history):
+        bug_ids = set()
+        keyword_history = self.create_keyword_map(history)
+
+        bugs_without_history = []
+        for bug in bugs:
+            bug_id = bug["id"]
+            current_keywords = bug["keywords"]
+            missing_keywords = []
+
+            # Check if keyword exists, but is not in "added" history
+            for keyword in current_keywords:
+                if bug_id not in keyword_history or (
+                    keyword not in keyword_history[bug_id]["added"]
+                ):
+                    if keyword not in missing_keywords:
+                        missing_keywords.append(keyword)
+
+            # Check for keywords that don't have "added" record,
+            # bug have "removed" record
+            if bug_id in keyword_history:
+                for keyword in keyword_history[bug_id]["removed"]:
+                    if (
+                        keyword not in keyword_history[bug_id]["added"]
+                        and keyword not in missing_keywords
+                    ):
+                        missing_keywords.append(keyword)
+
+            if missing_keywords:
+                bugs_without_history.append((bug, missing_keywords))
+
+        result = []
+
+        for _, (item, keywords) in enumerate(bugs_without_history):
+            record = {
+                "number": item["id"],
+                "who": item["creator"],
+                "change_time": item["creation_time"],
+                "changes": [
+                    {
+                        "added": ", ".join(keywords),
+                        "field_name": "keywords",
+                        "removed": "",
+                    }
+                ],
+            }
+            bug_ids.add(item["id"])
+            result.append(record)
+
+        return result
+
+    def fetch_history_for_new_bugs(self, all_bugs):
+        only_unsaved_changes = []
+
+        existing_ids = self.get_imported_ids()
+        all_ids = {bug["id"] for bug in all_bugs}
+        new_ids = all_ids - existing_ids
+
+        logging.info(f"Fetching new bugs history: {list(new_ids)}")
+
+        new_bugs = [bug for bug in all_bugs if bug["id"] in new_ids]
+
+        history, _ = self.extract_relevant_changes(self.fetch_bugs_history(new_ids))
+
+        synthetic_history = self.create_synthetic_history(new_bugs, history)
```

Comment:
synthetic_history is only created for new bugs. (I did a re import from scratch and saved all history in `bugs_history_copy`, will rename it to `bugs_history` once this is merged)